File: ui/src/lib/utils/validation.test.ts
import { describe, expect, it } from 'vitest';
import {
	PASSWORD_MAX_LENGTH,
	RESET_CODE_LENGTH,
	validatePassword,
	validateUsername
} from './validation';

describe('validatePassword', () => {
	it('rejects an empty password', () => {
		expect(validatePassword('').valid).toBe(false);
	});

	it('accepts a single-character password (warn, do not block)', () => {
		expect(validatePassword('p').valid).toBe(true);
	});

	it('accepts a password at the maximum length', () => {
		expect(validatePassword('a'.repeat(PASSWORD_MAX_LENGTH)).valid).toBe(true);
	});

	it('rejects a password one character above the maximum', () => {
		expect(validatePassword('a'.repeat(PASSWORD_MAX_LENGTH + 1)).valid).toBe(false);
	});
});

describe('validateUsername', () => {
	it('rejects an empty username', () => {
		expect(validateUsername('').valid).toBe(false);
	});

	it('accepts a username in range', () => {
		expect(validateUsername('alice').valid).toBe(true);
	});
});

describe('RESET_CODE_LENGTH', () => {
	it('is 8', () => {
		expect(RESET_CODE_LENGTH).toBe(8);
	});
});

File: server/internal/auth/validation_test.go
package auth

import "testing"

func TestValidatePassword(t *testing.T) {
	tests := []struct {
		name     string
		password string
		wantErr  bool
	}{
		{"empty", "", true},
		{"single character accepted (warn, don't block)", "p", false},
		{"in range", "a-normal-password", false},
		{"at maximum", makeString(MaxPasswordLength), false},
		{"one above maximum", makeString(MaxPasswordLength + 1), true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidatePassword(tt.password)
			if (err != nil) != tt.wantErr {
				t.Errorf("ValidatePassword(%q) error = %v, wantErr %v", tt.password, err, tt.wantErr)
			}
		})
	}
}

func TestValidateUsername(t *testing.T) {
	tests := []struct {
		name     string
		username string
		wantErr  bool
	}{
		{"empty", "", true},
		{"one below minimum", "a", true},
		{"at minimum", "ab", false},
		{"in range", "alice", false},
		{"at maximum", makeString(MaxUsernameLength), false},
		{"one above maximum", makeString(MaxUsernameLength + 1), true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidateUsername(tt.username)
			if (err != nil) != tt.wantErr {
				t.Errorf("ValidateUsername(%q) error = %v, wantErr %v", tt.username, err, tt.wantErr)
			}
		})
	}
}

func TestValidateResetCode(t *testing.T) {
	tests := []struct {
		name    string
		code    string
		wantErr bool
	}{
		{"empty", "", true},
		{"one short", "1234567", true},
		{"exact length", "12345678", false},
		{"one long", "123456789", true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidateResetCode(tt.code)
			if (err != nil) != tt.wantErr {
				t.Errorf("ValidateResetCode(%q) error = %v, wantErr %v", tt.code, err, tt.wantErr)
			}
		})
	}
}

func makeString(n int) string {
	b := make([]byte, n)
	for i := range b {
		b[i] = 'a'
	}
	return string(b)
}
